File: Linked List/2. Add Two Numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    struct ListNode* result = malloc(sizeof(struct ListNode)); 
    struct ListNode* ptr = result;
    result->val = 0;
    result->next = NULL;

    int carry = 0;
    
    while (l1 != NULL || l2!=NULL || carry !=0){
        
        int a = (l1 == NULL) ? 0 : l1->val;
        int b = (l2 == NULL) ? 0 : l2->val;
        ptr->val = a + b + carry;
        carry = ptr->val / 10;
        ptr->val = ptr->val %10;

        if(l1!=NULL){
            l1 = (l1->next ==NULL)? NULL: l1->next;
        }
        if(l2 !=NULL){
            l2 = (l2->next !=NULL)? l2->next: NULL;
        }

        if(l1 != NULL || l2!=NULL || carry !=0){
            ptr->next = malloc(sizeof(struct ListNode));
            ptr->next->next=NULL;
            ptr = ptr->next;
        }
        
    }
    return result;
}

File: Linked List/2. Add Two Numbers.py
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0 
        dummy = ListNode()
        curr = dummy
        
        while l1 or l2 or carry:
            
            v1 = l1.val if l1 else 0
            v2 = l2.val if l2 else 0 
            
            total = (v1+v2+carry)%10
            carry = (v1+v2+carry)//10
            curr.next = ListNode(total)
            curr = curr.next
            
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        
        return dummy.next
